ecommerce/extensions/payment/processors/payu.py

```diff
@@ -116,6 +121,21 @@ def get_transaction_parameters(self, basket, request=None, use_client_side_check
 
         return parameters
 
+    def apply_basket_vouchers_cleaning(self, basket, request):
+        """
+        Apply voucher basket cleaning to prevent submitting a basket with invalid vouchers
+        """
+        if basket.num_lines == 1 and basket.lines.first().product.is_enrollment_code_product:
+            basket.clear_vouchers()
+        elif basket.contains_a_voucher:
+            for voucher in basket.vouchers.all():
+                basket.vouchers.remove(voucher)
```

Comment:
You're calling remove for each voucher. Is that better than saving the vouchers in a list and then remove them from the QS?
https://docs.djangoproject.com/en/2.2/topics/db/optimization/#remove-in-bulk